src/runtime/hexagon_cache_allocator.cpp

```diff
@@ -0,0 +1,165 @@
+    halide_mutex_lock(&hexagon_cache_mutex);
+    pcache_pool temp = hexagon_cache_pool;
+    pcache_pool prev = NULL;
+
+    // Walk the free list for chunk of memory.
+    //Simplistic approach if free block size is less than the size requested assign
+    while (temp != NULL) {
+        if ((temp->used == false) &&
+            (size == temp->bytes)) {
+            temp->used = true;
+            halide_mutex_unlock(&hexagon_cache_mutex);
+            return (void *)temp->l2memory;
+        }
+        prev = temp;
+        temp = temp->next;
+    }
+
+    // If we are still here that means temp was null.
+    temp = (pcache_pool) malloc(sizeof(hexagon_cache_pool_t));
+    if (temp == NULL) {
+        halide_print(user_context, "Hexagon Cache Pool Allocation Failed.\n");
+        halide_mutex_unlock(&hexagon_cache_mutex);     
+        return NULL;
+    }
+    uint8_t *mem = (uint8_t *)HAP_cache_lock(sizeof(char) * size, NULL);
+    if ((mem == NULL) && retry) {
+        halide_print(user_context, "HAP_cache_lock failed, try deallocating unused cache.\n");
+        //Walk the list and deallocate unused blocks
+        pcache_pool temp2 = hexagon_cache_pool;
+        pcache_pool prev_node = hexagon_cache_pool;
+        while (temp2 != NULL) {
+            if (temp2->used == false) {
+                HAP_cache_unlock(temp2->l2memory);
+                //set previous node details
+                prev_node->next = (temp2->next)->next;
+                prev_node = temp2->next;
+                //set head
+                if (temp2 == hexagon_cache_pool) {
+                    hexagon_cache_pool = temp2->next;
+                 }
+                 //free node and reassign the variable
+                 free(temp2);
+                 temp2 = prev_node;
+                }
+                prev_node = temp2;
+                temp2 = temp2->next;
+        }
+        //Retry one more time after deallocating unused nodes
+        mem = (uint8_t *)HAP_cache_lock(sizeof(char) * size, NULL);
+        prev = prev_node;
+        if (mem == NULL) {
+            free(temp);
+            halide_print(user_context, "HAP_cache_lock failed.\n");
+            halide_mutex_unlock(&hexagon_cache_mutex);
+            return NULL;
+        }
+    } else if (mem == NULL) {
+        halide_print(user_context, "HAP_cache_lock failed.\n");
+        halide_mutex_unlock(&hexagon_cache_mutex);
+        return NULL;
+    }
+    temp->l2memory = (void *)mem;
+    temp->bytes = size;
+    temp->used = true;
+    temp->next = NULL;
+
+    if (prev != NULL) {
+        prev->next = temp;
+    } else if (hexagon_cache_pool == NULL) {
+        hexagon_cache_pool = temp;
+    }
+    halide_mutex_unlock(&hexagon_cache_mutex);
+    return (void*) temp->l2memory;
+}
+
+static inline void hexagon_cache_pool_put(void *user_context, void *cache_mem) {
+    halide_assert(user_context, cache_mem);
+    halide_mutex_lock(&hexagon_cache_mutex);
```

Comment:
Use ScopedMutexLock where possible: https://github.com/halide/Halide/blob/master/src/runtime/scoped_mutex_lock.h#L9

Reply:
Used ScopedMutexLock instead of halide_mutex_lock()